<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>cTRAP package — cTRAP • cTRAP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="cTRAP package — cTRAP"><meta property="og:description" content="Compare differential gene expression results with those from big datasets
(e.g. CMap), allowing to infer which types of perturbations may explain the
observed difference in gene expression.
Optimised to run in ShinyProxy with Celery/Flower backend with argument
shinyproxy = TRUE."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-2');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cTRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/cTRAP.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/nuno-agostinho/cTRAP/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>cTRAP package</h1>
    <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/cTRAP/blob/HEAD/R/cTRAP-package.r" class="external-link"><code>R/cTRAP-package.r</code></a>, <a href="https://github.com/nuno-agostinho/cTRAP/blob/HEAD/R/shinyInterface_session.R" class="external-link"><code>R/shinyInterface_session.R</code></a></small>
    <div class="hidden name"><code>cTRAP.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compare differential gene expression results with those from big datasets
(e.g. CMap), allowing to infer which types of perturbations may explain the
observed difference in gene expression.</p>
<p>Optimised to run in ShinyProxy with Celery/Flower backend with argument
<code>shinyproxy = TRUE</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">cTRAP</span><span class="op">(</span>
  <span class="va">...</span>,
  commonPath <span class="op">=</span> <span class="st">"data"</span>,
  expire <span class="op">=</span> <span class="fl">14</span>,
  fileSizeLimitMiB <span class="op">=</span> <span class="fl">50</span>,
  flowerURL <span class="op">=</span> <span class="cn">NULL</span>,
  port <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"shiny.port"</span><span class="op">)</span>,
  host <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"shiny.host"</span>, <span class="st">"127.0.0.1"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>...</dt>
<dd><p>Objects</p></dd>


<dt>commonPath</dt>
<dd><p>Character: path where to store data common to all sessions</p></dd>


<dt>expire</dt>
<dd><p>Character: days until a session expires (message purposes only)</p></dd>


<dt>fileSizeLimitMiB</dt>
<dd><p>Numeric: file size limit in MiB</p></dd>


<dt>flowerURL</dt>
<dd><p>Character: Flower REST API's URL (<code>NULL</code> to avoid using
Celery/Flower backend)</p></dd>


<dt>port</dt>
<dd><p>The TCP port that the application should listen on. If the
<code>port</code> is not specified, and the <code>shiny.port</code> option is set (with
<code>options(shiny.port = XX)</code>), then that port will be used. Otherwise,
use a random port between 3000:8000, excluding ports that are blocked
by Google Chrome for being considered unsafe: 3659, 4045, 5060,
5061, 6000, 6566, 6665:6669 and 6697. Up to twenty random
ports will be tried.</p></dd>


<dt>host</dt>
<dd><p>The IPv4 address that the application should listen on. Defaults
to the <code>shiny.host</code> option, if set, or <code>"127.0.0.1"</code> if not. See
Details.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Launches result viewer and plotter (returns <code>NULL</code>)</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Input:</strong> To use this package, a named vector of differentially
expressed gene metric is needed, where its values represent the significance
and magnitude of the differentially expressed genes (e.g. t-statistic) and
its names are gene symbols.</p>
<p><strong>Workflow:</strong> The differentially expressed genes will be compared
against selected perturbation conditions by:</p><ul><li><p>Spearman or Pearson correlation with z-scores of differentially
    expressed genes after perturbations from CMap. Use function
    <code><a href="rankSimilarPerturbations.html">rankSimilarPerturbations</a></code> with <code>method = "spearman"</code> or
    <code>method = "pearson"</code></p></li>
<li><p>Gene set enrichment analysis (GSEA) using the (around) 12 000 genes
    from CMap. Use function <code><a href="rankSimilarPerturbations.html">rankSimilarPerturbations</a></code> with
    <code>method = gsea</code>.</p></li>
</ul><p>Available perturbation conditions for CMap include:</p><ul><li><p>Cell line(s).</p></li>
<li><p>Perturbation type (gene knockdown, gene upregulation or drug
    intake).</p></li>
<li><p>Drug concentration.</p></li>
<li><p>Time points.</p></li>
</ul><p>Values for each perturbation type can be listed with
<code><a href="getCMapPerturbationTypes.html">getCMapPerturbationTypes()</a></code></p>
<p><strong>Output:</strong> The output includes a data frame of ranked perturbations
based on the associated statistical values and respective p-values.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other visual interface functions: 
<code><a href="launchCMapDataLoader.html">launchCMapDataLoader</a>()</code>,
<code><a href="launchDiffExprLoader.html">launchDiffExprLoader</a>()</code>,
<code><a href="launchDrugSetEnrichmentAnalyser.html">launchDrugSetEnrichmentAnalyser</a>()</code>,
<code><a href="launchMetadataViewer.html">launchMetadataViewer</a>()</code>,
<code><a href="launchResultPlotter.html">launchResultPlotter</a>()</code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernardo P. de Almeida, Nuno Saraiva-Agostinho, Nuno L. Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

