<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>cTRAP: identifying candidate causal perturbations from differential gene expression data • cTRAP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="cTRAP: identifying candidate causal perturbations from differential gene expression data">
<meta property="og:description" content="cTRAP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-2"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-2');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cTRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.10.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cTRAP.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nuno-agostinho/cTRAP/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cTRAP_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>cTRAP: identifying candidate causal perturbations from differential gene expression data</h1>
                        <h4 class="author">Bernardo P. de Almeida &amp; Nuno Saraiva-Agostinho</h4>
            
            <h4 class="date">2021-10-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nuno-agostinho/cTRAP/blob/master/vignettes/cTRAP.Rmd"><code>vignettes/cTRAP.Rmd</code></a></small>
      <div class="hidden name"><code>cTRAP.Rmd</code></div>

    </div>

    
    
<hr>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>cTRAP</code> is an R package designed to compare differential gene expression results with those from known cellular perturbations (such as gene knockdown, overexpression or small molecules) derived from the <a href="https://clue.io/">Connectivity Map</a> (CMap; <a href="https://doi.org/10.1016/j.cell.2017.10.049">Subramanian et al., Cell 2017</a>). Such analyses allow not only to infer the molecular causes of the observed difference in gene expression but also to identify small molecules that could drive or revert specific transcriptomic alterations.</p>
<p>To illustrate the package functionalities, we will use an example based on a gene knockdown dataset from the <a href="https://www.encodeproject.org/">ENCODE project</a> for which there is available RNA-seq data. After performing differential expression analyses to the matched-control sample, we will compare the respective transcriptomic changes with the ones caused by all CMap’s gene knockdown perturbations to identify which ones have similar or inverse transcriptomic changes to the observed ones. As a positive control, we expect to find the knock-down of the gene depleted in the ENCODE experiment as one of the most similar transcriptomic perturbations.</p>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<p>To load the <code>cTRAP</code> package into your R environment type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://nuno-agostinho.github.io/cTRAP">cTRAP</a></span><span class="op">)</span></code></pre></div>
<pre><code>## Registered S3 method overwritten by 'quantmod':
##   method            from
##   as.zoo.data.frame zoo</code></pre>
</div>
<div id="load-encode-rna-seq-data-and-perform-differential-gene-expression-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#load-encode-rna-seq-data-and-perform-differential-gene-expression-analysis" class="anchor"></a>Load ENCODE RNA-seq data and perform differential gene expression analysis</h1>
<p>In this example, we will use the EIF4G1 shRNA knockdown followed by RNA-seq experiment in HepG2 cell line from the ENCODE project as the dataset of interest. The RNA-seq processed data (gene quantifications from RSEM method) for the EIF4G1 knock-down and respective controls (two replicates each) can be automatically downloaded and loaded by typing:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"EIF4G1"</span>
<span class="va">cellLine</span> <span class="op">&lt;-</span> <span class="st">"HepG2"</span>

<span class="va">ENCODEmetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadENCODEknockdownMetadata.html">downloadENCODEknockdownMetadata</a></span><span class="op">(</span><span class="va">cellLine</span>, <span class="va">gene</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ENCODEmetadata</span><span class="op">$</span><span class="va">`Experiment target`</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">ENCODEmetadata</span><span class="op">$</span><span class="va">`Experiment target`</span><span class="op">)</span><span class="op">)</span>

<span class="va">ENCODEsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadENCODEsamples.html">loadENCODEsamples</a></span><span class="op">(</span><span class="va">ENCODEmetadata</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareENCODEgeneExpression.html">prepareENCODEgeneExpression</a></span><span class="op">(</span><span class="va">ENCODEsamples</span><span class="op">)</span></code></pre></div>
<p>Gene expression data (read counts) were quantile-normalized using <a href="https://www.ncbi.nlm.nih.gov/pubmed/24485249"><code>voom</code></a> and differential expression analysis was performed using the <a href="https://www.ncbi.nlm.nih.gov/pubmed/25605792"><code>limma</code></a> R package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Remove low coverage (at least 10 counts shared across two samples)</span>
<span class="va">minReads</span>   <span class="op">&lt;-</span> <span class="fl">10</span>
<span class="va">minSamples</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="va">filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span> , <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="va">minReads</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">minSamples</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="va">filter</span>, <span class="op">]</span>

<span class="co"># Convert ENSEMBL identifier to gene symbol</span>
<span class="va">counts</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convertENSEMBLtoGeneSymbols.html">convertENSEMBLtoGeneSymbols</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>

<span class="co"># Perform differential gene expression analysis</span>
<span class="va">diffExpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/performDifferentialExpression.html">performDifferentialExpression</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></code></pre></div>
<p>For our metric of differential expression after EIF4G1 shRNA knock-down, we will use the respective t-statistic.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get t-statistics of differential expression with respective gene names </span>
<span class="co"># (expected input for downstream analyses)</span>
<span class="va">diffExprStat</span> <span class="op">&lt;-</span> <span class="va">diffExpr</span><span class="op">$</span><span class="va">t</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">diffExprStat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">diffExpr</span><span class="op">$</span><span class="va">Gene_symbol</span></code></pre></div>
</div>
<div id="load-cmap-perturbation-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-cmap-perturbation-data" class="anchor"></a>Load CMap perturbation data</h1>
<p>We will use our differential gene expression metric to compare with CMap’s gene knock-down perturbations in the same cell line (HepG2). Note that this comparison can also be done to perturbations in a different cell line (or in all cell lines using the average result across cell lines).</p>
<p>To summarise conditions and check available data in CMap, we can use the following commands to download CMap metadata:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load CMap metadata (automatically downloaded if not found)</span>
<span class="va">cmapMetadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadCMapData.html">loadCMapData</a></span><span class="op">(</span><span class="st">"cmapMetadata.txt"</span>, type<span class="op">=</span><span class="st">"metadata"</span><span class="op">)</span></code></pre></div>
<pre><code>## cmapMetadata.txt not found: downloading data...</code></pre>
<pre><code>## Extracting cmapMetadata.txt.gz...</code></pre>
<pre><code>## Loading CMap metadata from cmapMetadata.txt...</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise conditions for all CMap perturbations</span>
<span class="fu"><a href="../reference/getCMapConditions.html">getCMapConditions</a></span><span class="op">(</span><span class="va">cmapMetadata</span><span class="op">)</span></code></pre></div>
<pre><code>## $perturbationType
## [1] "Compound"                                               
## [2] "Peptides and other biological agents (e.g. cytokine)"   
## [3] "shRNA for loss of function (LoF) of gene"               
## [4] "Consensus signature from shRNAs targeting the same gene"
## [5] "cDNA for overexpression of wild-type gene"              
## [6] "cDNA for overexpression of mutated gene"                
## 
## $cellLine
##  [1] "A375"     "A549"     "A673"     "AGS"      "ASC"      "BT20"    
##  [7] "CD34"     "CL34"     "CORL23"   "COV644"   "DV90"     "EFO27"   
## [13] "FIBRNPC"  "H1299"    "HA1E"     "HCC15"    "HCC515"   "HCT116"  
## [19] "HEC108"   "HEK293T"  "HEKTE"    "HEPG2"    "HL60"     "HS27A"   
## [25] "HS578T"   "HT115"    "HT29"     "HUH7"     "JHUEM2"   "JURKAT"  
## [31] "LOVO"     "MCF10A"   "MCF7"     "MCH58"    "MDAMB231" "MDST8"   
## [37] "NCIH1694" "NCIH1836" "NCIH2073" "NCIH508"  "NCIH596"  "NCIH716" 
## [43] "NEU"      "NKDBA"    "NOMO1"    "NPC"      "OV7"      "PC3"     
## [49] "PHH"      "PL21"     "RKO"      "RMGI"     "RMUGS"    "SHSY5Y"  
## [55] "SKB"      "SKBR3"    "SKL"      "SKLU1"    "SKM1"     "SKMEL1"  
## [61] "SKMEL28"  "SNGM"     "SNU1040"  "SNUC4"    "SNUC5"    "SW480"   
## [67] "SW620"    "SW948"    "T3M10"    "THP1"     "TYKNU"    "U266"    
## [73] "U2OS"     "U937"     "VCAP"     "WSUDLCL2"
## 
## $dosage
##  [1] NA             "0.1 %"        "1 nM"         "10 nM"        "100 nM"      
##  [6] "500 nM"       "0.04 µM"      "0.12 µM"      "0.37 µM"      "0.41 µM"     
## [11] "1 µM"         "1.11 µM"      "1.23 µM"      "3 µM"         "3.33 µM"     
## [16] "3.7 µM"       "5 µM"         "10 µM"        "11.11 µM"     "20 µM"       
## [21] "30 µM"        "33.33 µM"     "40 µM"        "50 µM"        "60 µM"       
## [26] "70 µM"        "80 µM"        "90 µM"        "100 µM"       "1 µL"        
## [31] "1.5 µL"       "2 µL"         "3 µL"         "4 µL"         "5 µL"        
## [36] "6 µL"         "10 µL"        "20 µL"        "150 ng"       "200 ng"      
## [41] "300 ng"       "0.1 ng/µL"    "1 ng/µL"      "3 ng/µL"      "10 ng/µL"    
## [46] "100 ng/µL"    "300 ng/µL"    "0 ng/mL"      "0.01 ng/mL"   "0.03 ng/mL"  
## [51] "0.1 ng/mL"    "0.15 ng/mL"   "0.2 ng/mL"    "0.3 ng/mL"    "0.5 ng/mL"   
## [56] "1 ng/mL"      "1.65 ng/mL"   "2 ng/mL"      "5 ng/mL"      "10 ng/mL"    
## [61] "15 ng/mL"     "16 ng/mL"     "20 ng/mL"     "25 ng/mL"     "30 ng/mL"    
## [66] "40 ng/mL"     "45 ng/mL"     "50 ng/mL"     "80 ng/mL"     "100 ng/mL"   
## [71] "150 ng/mL"    "200 ng/mL"    "250 ng/mL"    "400 ng/mL"    "500 ng/mL"   
## [76] "800 ng/mL"    "1000 ng/mL"   "2000 ng/mL"   "2500 ng/mL"   "3000 ng/mL"  
## [81] "5000 ng/mL"   "8300 ng/mL"   "10000 ng/mL"  "50000 ng/mL"  "100000 ng/mL"
## [86] "200000 ng/mL"
## 
## $timepoint
##  [1] "1 h"   "2 h"   "3 h"   "4 h"   "6 h"   "24 h"  "48 h"  "72 h"  "96 h" 
## [10] "120 h" "144 h" "168 h"</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise conditions for CMap perturbations in HepG2 cell line</span>
<span class="fu"><a href="../reference/getCMapConditions.html">getCMapConditions</a></span><span class="op">(</span><span class="va">cmapMetadata</span>, cellLine<span class="op">=</span><span class="st">"HepG2"</span><span class="op">)</span></code></pre></div>
<pre><code>## $perturbationType
## [1] "Compound"                                               
## [2] "Peptides and other biological agents (e.g. cytokine)"   
## [3] "shRNA for loss of function (LoF) of gene"               
## [4] "Consensus signature from shRNAs targeting the same gene"
## [5] "cDNA for overexpression of wild-type gene"              
## 
## $cellLine
## [1] "HEPG2"
## 
## $dosage
##  [1] NA             "0.1 %"        "10 nM"        "100 nM"       "500 nM"      
##  [6] "0.04 µM"      "0.12 µM"      "0.37 µM"      "0.41 µM"      "1 µM"        
## [11] "1.11 µM"      "1.23 µM"      "3 µM"         "3.33 µM"      "3.7 µM"      
## [16] "5 µM"         "10 µM"        "11.11 µM"     "20 µM"        "30 µM"       
## [21] "33.33 µM"     "40 µM"        "50 µM"        "60 µM"        "70 µM"       
## [26] "80 µM"        "90 µM"        "100 µM"       "1.5 µL"       "2 µL"        
## [31] "0 ng/mL"      "0.01 ng/mL"   "0.03 ng/mL"   "0.1 ng/mL"    "0.15 ng/mL"  
## [36] "0.2 ng/mL"    "0.3 ng/mL"    "0.5 ng/mL"    "1 ng/mL"      "1.65 ng/mL"  
## [41] "2 ng/mL"      "5 ng/mL"      "10 ng/mL"     "15 ng/mL"     "16 ng/mL"    
## [46] "20 ng/mL"     "25 ng/mL"     "30 ng/mL"     "40 ng/mL"     "45 ng/mL"    
## [51] "50 ng/mL"     "80 ng/mL"     "100 ng/mL"    "150 ng/mL"    "200 ng/mL"   
## [56] "250 ng/mL"    "400 ng/mL"    "500 ng/mL"    "800 ng/mL"    "1000 ng/mL"  
## [61] "2000 ng/mL"   "2500 ng/mL"   "3000 ng/mL"   "5000 ng/mL"   "8300 ng/mL"  
## [66] "10000 ng/mL"  "50000 ng/mL"  "100000 ng/mL" "200000 ng/mL"
## 
## $timepoint
## [1] "2 h"  "4 h"  "6 h"  "24 h" "96 h"</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarise conditions for a specific CMap perturbation in HepG2 cell line</span>
<span class="fu"><a href="../reference/getCMapConditions.html">getCMapConditions</a></span><span class="op">(</span>
    <span class="va">cmapMetadata</span>, cellLine<span class="op">=</span><span class="st">"HepG2"</span>,
    perturbationType<span class="op">=</span><span class="st">"Consensus signature from shRNAs targeting the same gene"</span><span class="op">)</span></code></pre></div>
<pre><code>## $perturbationType
## [1] "Consensus signature from shRNAs targeting the same gene"
## 
## $cellLine
## [1] "HEPG2"
## 
## $dosage
## [1] "1.5 µL"
## 
## $timepoint
## [1] "96 h"</code></pre>
<p>Now, we will filter the metadata to CMap gene knockdown perturbations in HepG2 cell line and load associated gene information and differential gene expression data. Data is loaded based on the given filename (if no file with the given filename is available, data will be first downloaded).</p>
<blockquote>
<p>Differential gene expression data for each CMap perturbation are available in normalised z-score values (read <a href="https://doi.org/10.1016/j.cell.2017.10.049">Subramanian et al., Cell 2017</a> for more details). As this file’s size is big (around 20GB), a prompt will ask you to confirm whether to start downloading the file.</p>
<p>Alternatively, you can download the file from the following link (do not forget to use the same filename as the argument of <code><a href="../reference/prepareCMapPerturbations.html">prepareCMapPerturbations()</a></code>): <a href="ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx.gz" class="uri">ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx.gz</a></p>
</blockquote>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filter CMap gene knockdown HepG2 data to be loaded</span>
<span class="va">cmapMetadataKD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterCMapMetadata.html">filterCMapMetadata</a></span><span class="op">(</span>
    <span class="va">cmapMetadata</span>, cellLine<span class="op">=</span><span class="st">"HepG2"</span>,
    perturbationType<span class="op">=</span><span class="st">"Consensus signature from shRNAs targeting the same gene"</span><span class="op">)</span>

<span class="co"># Load filtered data (data will be downloaded if not found)</span>
<span class="va">cmapPerturbationsKD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareCMapPerturbations.html">prepareCMapPerturbations</a></span><span class="op">(</span>
    metadata<span class="op">=</span><span class="va">cmapMetadataKD</span>, zscores<span class="op">=</span><span class="st">"cmapZscores.gctx"</span>,
    geneInfo<span class="op">=</span><span class="st">"cmapGeneInfo.txt"</span><span class="op">)</span></code></pre></div>
<p>If interested in small molecules, the differential gene expression z-scores from CMap can be downloaded for each small molecule perturbation:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filter CMap gene small molecule HepG2 data to be loaded</span>
<span class="va">cmapMetadataCompounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterCMapMetadata.html">filterCMapMetadata</a></span><span class="op">(</span>
    <span class="va">cmapMetadata</span>, cellLine<span class="op">=</span><span class="st">"HepG2"</span>, perturbationType<span class="op">=</span><span class="st">"Compound"</span><span class="op">)</span>

<span class="co"># Load filtered data (data will be downloaded if not found)</span>
<span class="va">cmapPerturbationsCompounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareCMapPerturbations.html">prepareCMapPerturbations</a></span><span class="op">(</span>
    metadata<span class="op">=</span><span class="va">cmapMetadataCompounds</span>, zscores<span class="op">=</span><span class="st">"cmapZscores.gctx"</span>,
    geneInfo<span class="op">=</span><span class="st">"cmapGeneInfo.txt"</span>, compoundInfo<span class="op">=</span><span class="st">"cmapCompoundInfo.txt"</span><span class="op">)</span></code></pre></div>
</div>
<div id="comparison-with-cmap-perturbations" class="section level1">
<h1 class="hasAnchor">
<a href="#comparison-with-cmap-perturbations" class="anchor"></a>Comparison with CMap perturbations</h1>
<p>The <code>rankSimilarPerturbations</code> function compares the differential expression metric (the t-statistic, in this case) against the CMap perturbations’ z-scores using the available methods:</p>
<ul>
<li>Spearman’s correlation</li>
<li>Pearson’s correlation</li>
<li>Gene Set Enrichment Analysis (GSEA), where the most up- and down-regulated <em>n</em> genes from the user’s differential expression profile are used as gene sets (by default, <em>n</em> = 150 genes)</li>
</ul>
<p>To compare against CMap knockdown perturbations using all the previous methods:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">compareKD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rankSimilarPerturbations.html">rankSimilarPerturbations</a></span><span class="op">(</span><span class="va">diffExprStat</span>, <span class="va">cmapPerturbationsKD</span><span class="op">)</span></code></pre></div>
<p>To compare against selected CMap small molecule perturbations:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">compareCompounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rankSimilarPerturbations.html">rankSimilarPerturbations</a></span><span class="op">(</span><span class="va">diffExprStat</span>, 
                                             <span class="va">cmapPerturbationsCompounds</span><span class="op">)</span></code></pre></div>
<p>The output table contains the results of the comparison with each perturbation tested, including the test statistics (Spearman’s correlation coefficient, Pearson’s correlation coefficient and/or GSEA score), the respective p-value and the Benjamini-Hochberg-adjusted p-value (for correlation statistics only). When performing multiple methods, the <a href="https://en.wikipedia.org/wiki/Rank_product">rank product</a>’s rank will be included to summarise other method’s rankings.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Most positively associated perturbations (note that EIF4G1 knockdown is the</span>
<span class="co"># 7th, 1st and 2nd most positively associated perturbation based on Spearman's</span>
<span class="co"># correlation, Pearson's correlation and GSEA, respectively)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">spearman_rank</span><span class="op">)</span><span class="op">]</span>, n<span class="op">=</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>##                 gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
##  1:    CGS001_HEPG2_96H:MECP2:1.5     0.1930335    1.506067e-74    1.305258e-73
##  2: CGS001_HEPG2_96H:KIAA0196:1.5     0.1894301    8.237226e-72    5.354197e-71
##  3:    CGS001_HEPG2_96H:SQRDL:1.5     0.1883210    5.589838e-71    2.906716e-70
##  4:     CGS001_HEPG2_96H:PPIH:1.5     0.1863954    1.509735e-69    6.542184e-69
##  5:    CGS001_HEPG2_96H:STAT1:1.5     0.1833382    2.626628e-67    9.756048e-67
##  6:    CGS001_HEPG2_96H:COPS5:1.5     0.1775880    3.364592e-63    1.093493e-62
##  7:   CGS001_HEPG2_96H:EIF4G1:1.5     0.1770993    7.406639e-63    2.139696e-62
##  8:     CGS001_HEPG2_96H:MEST:1.5     0.1726479    8.816544e-60    2.292301e-59
##  9:   CGS001_HEPG2_96H:KIF20A:1.5     0.1665159    1.114839e-55    2.415485e-55
## 10:    CGS001_HEPG2_96H:TMEM5:1.5     0.1506335    8.936446e-46    1.452172e-45
##     pearson_coef pearson_pvalue pearson_qvalue      GSEA spearman_rank
##  1:    0.1756788   7.245930e-62   1.883942e-61 0.3903862             1
##  2:    0.1847352   2.514633e-68   9.340065e-68 0.3521034             2
##  3:    0.1718559   3.046859e-59   6.093718e-59 0.3998024             3
##  4:    0.1703630   3.104268e-58   5.380731e-58 0.4359134             4
##  5:    0.1894661   7.739198e-72   4.024383e-71 0.3765351             5
##  6:    0.1794790   1.553956e-64   5.050356e-64 0.4222756             6
##  7:    0.1915142   2.182595e-73   1.418686e-72 0.4861485             7
##  8:    0.1858475   3.830708e-69   1.659973e-68 0.4219693             8
##  9:    0.1786967   5.568601e-64   1.608707e-63 0.4566918             9
## 10:    0.1670226   5.177976e-56   8.414211e-56 0.4721782            10
##     pearson_rank GSEA_rank rankProduct_rank
##  1:            7        12                2
##  2:            4        14                4
##  3:            9        11                8
##  4:           10         8                9
##  5:            2        13                5
##  6:            5         9                7
##  7:            1         2                1
##  8:            3        10                6
##  9:            6         7               10
## 10:           11         6               12</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">pearson_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##                gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1:   CGS001_HEPG2_96H:EIF4G1:1.5     0.1770993    7.406639e-63    2.139696e-62
## 2:    CGS001_HEPG2_96H:STAT1:1.5     0.1833382    2.626628e-67    9.756048e-67
## 3:     CGS001_HEPG2_96H:MEST:1.5     0.1726479    8.816544e-60    2.292301e-59
## 4: CGS001_HEPG2_96H:KIAA0196:1.5     0.1894301    8.237226e-72    5.354197e-71
## 5:    CGS001_HEPG2_96H:COPS5:1.5     0.1775880    3.364592e-63    1.093493e-62
## 6:   CGS001_HEPG2_96H:KIF20A:1.5     0.1665159    1.114839e-55    2.415485e-55
##    pearson_coef pearson_pvalue pearson_qvalue      GSEA spearman_rank
## 1:    0.1915142   2.182595e-73   1.418686e-72 0.4861485             7
## 2:    0.1894661   7.739198e-72   4.024383e-71 0.3765351             5
## 3:    0.1858475   3.830708e-69   1.659973e-68 0.4219693             8
## 4:    0.1847352   2.514633e-68   9.340065e-68 0.3521034             2
## 5:    0.1794790   1.553956e-64   5.050356e-64 0.4222756             6
## 6:    0.1786967   5.568601e-64   1.608707e-63 0.4566918             9
##    pearson_rank GSEA_rank rankProduct_rank
## 1:            1         2                1
## 2:            2        13                5
## 3:            3        10                6
## 4:            4        14                4
## 5:            5         9                7
## 6:            6         7               10</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">GSEA_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##              gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1: CGS001_HEPG2_96H:SKIV2L:1.5     0.1423207    5.305914e-41    8.114927e-41
## 2: CGS001_HEPG2_96H:EIF4G1:1.5     0.1770993    7.406639e-63    2.139696e-62
## 3: CGS001_HEPG2_96H:GTPBP8:1.5     0.1190220    4.210973e-29    4.760230e-29
## 4:    CGS001_HEPG2_96H:HFE:1.5     0.1235293    3.098180e-31    3.661486e-31
## 5: CGS001_HEPG2_96H:UBAP2L:1.5     0.1309546    6.320554e-35    8.216721e-35
## 6:  CGS001_HEPG2_96H:TMEM5:1.5     0.1506335    8.936446e-46    1.452172e-45
##    pearson_coef pearson_pvalue pearson_qvalue      GSEA spearman_rank
## 1:    0.1755377   9.078490e-62   2.145825e-61 0.5070024            11
## 2:    0.1915142   2.182595e-73   1.418686e-72 0.4861485             7
## 3:    0.1390818   3.229116e-39   3.816228e-39 0.4813972            14
## 4:    0.1499527   2.252417e-45   2.788707e-45 0.4813046            13
## 5:    0.1517482   1.948821e-46   2.533468e-46 0.4722054            12
## 6:    0.1670226   5.177976e-56   8.414211e-56 0.4721782            10
##    pearson_rank GSEA_rank rankProduct_rank
## 1:            8         1                3
## 2:            1         2                1
## 3:           14         3               11
## 4:           13         4               13
## 5:           12         5               14
## 6:           11         6               12</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">rankProduct_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##                gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1:   CGS001_HEPG2_96H:EIF4G1:1.5     0.1770993    7.406639e-63    2.139696e-62
## 2:    CGS001_HEPG2_96H:MECP2:1.5     0.1930335    1.506067e-74    1.305258e-73
## 3:   CGS001_HEPG2_96H:SKIV2L:1.5     0.1423207    5.305914e-41    8.114927e-41
## 4: CGS001_HEPG2_96H:KIAA0196:1.5     0.1894301    8.237226e-72    5.354197e-71
## 5:    CGS001_HEPG2_96H:STAT1:1.5     0.1833382    2.626628e-67    9.756048e-67
## 6:     CGS001_HEPG2_96H:MEST:1.5     0.1726479    8.816544e-60    2.292301e-59
##    pearson_coef pearson_pvalue pearson_qvalue      GSEA spearman_rank
## 1:    0.1915142   2.182595e-73   1.418686e-72 0.4861485             7
## 2:    0.1756788   7.245930e-62   1.883942e-61 0.3903862             1
## 3:    0.1755377   9.078490e-62   2.145825e-61 0.5070024            11
## 4:    0.1847352   2.514633e-68   9.340065e-68 0.3521034             2
## 5:    0.1894661   7.739198e-72   4.024383e-71 0.3765351             5
## 6:    0.1858475   3.830708e-69   1.659973e-68 0.4219693             8
##    pearson_rank GSEA_rank rankProduct_rank
## 1:            1         2                1
## 2:            7        12                2
## 3:            8         1                3
## 4:            4        14                4
## 5:            2        13                5
## 6:            3        10                6</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Most negatively associated perturbations</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">spearman_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##              gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1:   CGS001_HEPG2_96H:EYA1:1.5    -0.1993531    1.747554e-79    4.543641e-78
## 2:  CGS001_HEPG2_96H:CDCA8:1.5    -0.1941862    1.951544e-75    2.537007e-74
## 3: CGS001_HEPG2_96H:NDUFB6:1.5    -0.1677880    1.618399e-56    3.825307e-56
## 4:  CGS001_HEPG2_96H:EHMT2:1.5    -0.1640525    4.479303e-54    8.958605e-54
## 5: CGS001_HEPG2_96H:ZBTB24:1.5    -0.1632375    1.500986e-53    2.787546e-53
## 6:    CGS001_HEPG2_96H:MAF:1.5    -0.1581847    2.351511e-50    4.075952e-50
##    pearson_coef pearson_pvalue pearson_qvalue       GSEA spearman_rank
## 1:   -0.1952303   3.030978e-76   7.880542e-75 -0.3776731            26
## 2:   -0.1926944   2.740646e-74   2.375227e-73 -0.4739211            25
## 3:   -0.1646057   1.964288e-54   3.004205e-54  0.0000000            24
## 4:   -0.1590654   6.637816e-51   9.587957e-51 -0.3716818            23
## 5:   -0.1948077   6.448588e-76   8.383165e-75  0.0000000            22
## 6:   -0.1717844   3.406704e-59   6.326737e-59  0.0000000            21
##    pearson_rank GSEA_rank rankProduct_rank
## 1:           26        19               25
## 2:           24        26               26
## 3:           21        16               20
## 4:           20        18               21
## 5:           25        16               22
## 6:           22        16               19</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">pearson_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##               gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1:    CGS001_HEPG2_96H:EYA1:1.5    -0.1993531    1.747554e-79    4.543641e-78
## 2:  CGS001_HEPG2_96H:ZBTB24:1.5    -0.1632375    1.500986e-53    2.787546e-53
## 3:   CGS001_HEPG2_96H:CDCA8:1.5    -0.1941862    1.951544e-75    2.537007e-74
## 4: CGS001_HEPG2_96H:SULT1A2:1.5    -0.1414914    1.533416e-40    2.214934e-40
## 5:     CGS001_HEPG2_96H:MAF:1.5    -0.1581847    2.351511e-50    4.075952e-50
## 6:  CGS001_HEPG2_96H:NDUFB6:1.5    -0.1677880    1.618399e-56    3.825307e-56
##    pearson_coef pearson_pvalue pearson_qvalue       GSEA spearman_rank
## 1:   -0.1952303   3.030978e-76   7.880542e-75 -0.3776731            26
## 2:   -0.1948077   6.448588e-76   8.383165e-75  0.0000000            22
## 3:   -0.1926944   2.740646e-74   2.375227e-73 -0.4739211            25
## 4:   -0.1742848   6.666242e-61   1.444352e-60 -0.4086520            20
## 5:   -0.1717844   3.406704e-59   6.326737e-59  0.0000000            21
## 6:   -0.1646057   1.964288e-54   3.004205e-54  0.0000000            24
##    pearson_rank GSEA_rank rankProduct_rank
## 1:           26        19               25
## 2:           25        16               22
## 3:           24        26               26
## 4:           23        20               24
## 5:           22        16               19
## 6:           21        16               20</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">GSEA_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##             gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1: CGS001_HEPG2_96H:CDCA8:1.5   -0.19418619    1.951544e-75    2.537007e-74
## 2: CGS001_HEPG2_96H:DHX16:1.5   -0.14109904    2.527811e-40    3.459110e-40
## 3: CGS001_HEPG2_96H:SIAH2:1.5   -0.07825242    2.030400e-13    2.111616e-13
## 4: CGS001_HEPG2_96H:GNAI2:1.5   -0.05956052    2.293513e-08    2.293513e-08
## 5:   CGS001_HEPG2_96H:SHB:1.5   -0.11094246    1.768452e-25    1.915823e-25
## 6: CGS001_HEPG2_96H:PLOD2:1.5   -0.12393601    1.970762e-31    2.439991e-31
##    pearson_coef pearson_pvalue pearson_qvalue       GSEA spearman_rank
## 1:  -0.19269443   2.740646e-74   2.375227e-73 -0.4739211            25
## 2:  -0.15642416   2.883756e-49   3.946193e-49 -0.4379221            19
## 3:  -0.11367900   1.119976e-26   1.164775e-26 -0.4224346            16
## 4:  -0.08923031   5.233713e-17   5.233713e-17 -0.4166373            15
## 5:  -0.12875045   8.300805e-34   8.992538e-34 -0.4131576            17
## 6:  -0.13259151   9.070613e-36   1.025374e-35 -0.4130739            18
##    pearson_rank GSEA_rank rankProduct_rank
## 1:           24        26               26
## 2:           19        25               23
## 3:           16        24               16
## 4:           15        23               15
## 5:           17        22               17
## 6:           18        21               18</code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">rankProduct_rank</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##               gene_perturbation spearman_coef spearman_pvalue spearman_qvalue
## 1:   CGS001_HEPG2_96H:CDCA8:1.5    -0.1941862    1.951544e-75    2.537007e-74
## 2:    CGS001_HEPG2_96H:EYA1:1.5    -0.1993531    1.747554e-79    4.543641e-78
## 3: CGS001_HEPG2_96H:SULT1A2:1.5    -0.1414914    1.533416e-40    2.214934e-40
## 4:   CGS001_HEPG2_96H:DHX16:1.5    -0.1410990    2.527811e-40    3.459110e-40
## 5:  CGS001_HEPG2_96H:ZBTB24:1.5    -0.1632375    1.500986e-53    2.787546e-53
## 6:   CGS001_HEPG2_96H:EHMT2:1.5    -0.1640525    4.479303e-54    8.958605e-54
##    pearson_coef pearson_pvalue pearson_qvalue       GSEA spearman_rank
## 1:   -0.1926944   2.740646e-74   2.375227e-73 -0.4739211            25
## 2:   -0.1952303   3.030978e-76   7.880542e-75 -0.3776731            26
## 3:   -0.1742848   6.666242e-61   1.444352e-60 -0.4086520            20
## 4:   -0.1564242   2.883756e-49   3.946193e-49 -0.4379221            19
## 5:   -0.1948077   6.448588e-76   8.383165e-75  0.0000000            22
## 6:   -0.1590654   6.637816e-51   9.587957e-51 -0.3716818            23
##    pearson_rank GSEA_rank rankProduct_rank
## 1:           24        26               26
## 2:           26        19               25
## 3:           23        20               24
## 4:           19        25               23
## 5:           25        16               22
## 6:           20        18               21</code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot list of compared pertubations</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"spearman"</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/order%20knockdown%20perturbation%20comparison-1.png" width="576"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"pearson"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/order%20knockdown%20perturbation%20comparison-2.png" width="576"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/order%20knockdown%20perturbation%20comparison-3.png" width="576"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"rankProduct"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/order%20knockdown%20perturbation%20comparison-4.png" width="576"></p>
<p>For small molecules:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Most positively and negatively associated perturbations</span>
<span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">rankProduct_rank</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##                              compound_perturbation spearman_coef
##  1:    CVD001_HEPG2_24H:BRD-A14014306-001-01-1:4.1   0.225335666
##  2: CVD001_HEPG2_24H:BRD-K84595254-001-03-0:4.9444   0.085133810
##  3: CVD001_HEPG2_24H:BRD-K62810658-001-05-6:4.6768  -0.050201852
##  4:    CVD001_HEPG2_24H:BRD-K41172353-001-01-4:4.7   0.005161915
##  5: CVD001_HEPG2_24H:BRD-K60476892-001-02-1:4.1072  -0.077832390
##  6: CVD001_HEPG2_24H:BRD-K96188950-001-04-5:4.3967  -0.037174082
##  7:  CVD001_HEPG2_24H:BRD-K84389640-001-01-5:4.225  -0.078331814
##  8:  CVD001_HEPG2_24H:BRD-K77508012-001-01-9:6.025  -0.094208866
##  9:   CVD001_HEPG2_24H:BRD-K31030218-001-01-1:4.25  -0.125531451
## 10:    CVD001_HEPG2_24H:BRD-K94818765-001-01-0:4.8  -0.113418048
## 11:   CVD001_HEPG2_24H:BRD-A65142661-034-01-8:5.35  -0.164686339
##     spearman_pvalue spearman_qvalue pearson_coef pearson_pvalue pearson_qvalue
##  1:   1.361508e-101   1.497659e-100   0.22640668  1.442316e-102  1.586548e-101
##  2:    1.294659e-15    2.373541e-15   0.06834399   1.415829e-10   2.224875e-10
##  3:    2.489663e-06    3.042921e-06  -0.02551843   1.673267e-02   1.840594e-02
##  4:    6.284641e-01    6.284641e-01  -0.01267514   2.347404e-01   2.347404e-01
##  5:    2.727173e-13    3.749863e-13  -0.06254059   4.400100e-09   6.050137e-09
##  6:    4.903648e-04    5.394013e-04  -0.04580625   1.737439e-05   2.123537e-05
##  7:    1.919930e-13    3.017033e-13  -0.08264025   8.485519e-15   1.555679e-14
##  8:    8.674931e-19    1.908485e-18  -0.08283071   7.364690e-15   1.555679e-14
##  9:    3.293334e-32    1.207556e-31  -0.12054061   8.215231e-30   3.012251e-29
## 10:    1.461350e-26    4.018714e-26  -0.11053910   2.640853e-25   7.262345e-25
## 11:    1.741438e-54    9.577908e-54  -0.14797060   3.242185e-44   1.783202e-43
##           GSEA spearman_rank pearson_rank GSEA_rank rankProduct_rank
##  1:  0.4932836             1            1       1.0                1
##  2:  0.2179143             2            2       3.0                2
##  3:  0.2304579             5            4       2.0                3
##  4:  0.0000000             3            3       5.5                4
##  5:  0.0000000             6            6       5.5                5
##  6: -0.2821337             4            5      10.0                6
##  7:  0.0000000             7            7       5.5                7
##  8: -0.2442628             8            8       8.0                8
##  9:  0.0000000            10           10       5.5                9
## 10: -0.2621164             9            9       9.0               10
## 11: -0.3381341            11           11      11.0               11</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="st">"rankProduct"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/order%20small%20molecule%20perturbation%20comparison-1.png" width="576"></p>
<blockquote>
<p>The Gene Set Enrichment Analysis (GSEA) score is based on the Weighted Connectivity Score (WTCS), a composite and bi-directional version of the weighted Kolmogorov-Smirnov enrichment statistic (ES) (<a href="https://doi.org/10.1016/j.cell.2017.10.049">Subramanian et al., Cell 2017</a>).</p>
</blockquote>
<blockquote>
<p>To calculate the GSEA score, GSEA is run for the most up- and down-regulated genes from the user’s differential expression profile. The GSEA score is the mean between ES<sub>top</sub> and ES<sub>bottom</sub> (however, if ES<sub>top</sub> and ES<sub>bottom</sub> have the same sign, the GSEA score is 0).</p>
</blockquote>
<blockquote>
<p>If a perturbation has a similar differential expression profile to our data (higher GSEA score), we expect to see the most up-regulated (down-regulated) genes in the perturbation enriched in the top (bottom) <em>n</em> differentially expressed genes from our data.</p>
</blockquote>
</div>
<div id="information-on-perturbations" class="section level1">
<h1 class="hasAnchor">
<a href="#information-on-perturbations" class="anchor"></a>Information on perturbations</h1>
<p>To get associated information as made available from CMap:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Information on the EIF4G1 knockdown perturbation</span>
<span class="va">EIF4G1knockdown</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"EIF4G1"</span>, <span class="va">compareKD</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, value<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">EIF4G1knockdown</span><span class="op">)</span></code></pre></div>
<pre><code>## $metadata
##                         sig_id     pert_id pert_iname  pert_type cell_id
## 1: CGS001_HEPG2_96H:EIF4G1:1.5 CGS001-1981     EIF4G1 trt_sh.cgs   HEPG2
##    pert_dose pert_dose_unit pert_idose pert_time pert_time_unit pert_itime
## 1:       1.5         microL 1.5 microL        96              h       96 h
##                                                                                                         distil_id
## 1: KDA004_HEPG2_96H:TRCN0000061769:-666|KDA004_HEPG2_96H:TRCN0000061770:-666|KDA004_HEPG2_96H:TRCN0000061772:-666</code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Information on the top 10 most similar compound perturbations (based on</span>
<span class="co"># Spearman's correlation)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">rankProduct_rank</span><span class="op">)</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## $metadata
##                            sig_id      pert_id pert_iname  pert_type cell_id
##  1:    CGS001_HEPG2_96H:COPS5:1.5 CGS001-10987      COPS5 trt_sh.cgs   HEPG2
##  2:   CGS001_HEPG2_96H:EIF4G1:1.5  CGS001-1981     EIF4G1 trt_sh.cgs   HEPG2
##  3: CGS001_HEPG2_96H:KIAA0196:1.5  CGS001-9897   KIAA0196 trt_sh.cgs   HEPG2
##  4:   CGS001_HEPG2_96H:KIF20A:1.5 CGS001-10112     KIF20A trt_sh.cgs   HEPG2
##  5:    CGS001_HEPG2_96H:MECP2:1.5  CGS001-4204      MECP2 trt_sh.cgs   HEPG2
##  6:     CGS001_HEPG2_96H:MEST:1.5  CGS001-4232       MEST trt_sh.cgs   HEPG2
##  7:     CGS001_HEPG2_96H:PPIH:1.5 CGS001-10465       PPIH trt_sh.cgs   HEPG2
##  8:   CGS001_HEPG2_96H:SKIV2L:1.5  CGS001-6499     SKIV2L trt_sh.cgs   HEPG2
##  9:    CGS001_HEPG2_96H:SQRDL:1.5 CGS001-58472      SQRDL trt_sh.cgs   HEPG2
## 10:    CGS001_HEPG2_96H:STAT1:1.5  CGS001-6772      STAT1 trt_sh.cgs   HEPG2
##     pert_dose pert_dose_unit pert_idose pert_time pert_time_unit pert_itime
##  1:       1.5         microL 1.5 microL        96              h       96 h
##  2:       1.5         microL 1.5 microL        96              h       96 h
##  3:       1.5         microL 1.5 microL        96              h       96 h
##  4:       1.5         microL 1.5 microL        96              h       96 h
##  5:       1.5         microL 1.5 microL        96              h       96 h
##  6:       1.5         microL 1.5 microL        96              h       96 h
##  7:       1.5         microL 1.5 microL        96              h       96 h
##  8:       1.5         microL 1.5 microL        96              h       96 h
##  9:       1.5         microL 1.5 microL        96              h       96 h
## 10:       1.5         microL 1.5 microL        96              h       96 h
##                                                                                                                                                                                                                                                                                                   distil_id
##  1:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000019200:-666|KDB009_HEPG2_96H:TRCN0000019201:-666|KDB009_HEPG2_96H:TRCN0000019202:-666
##  2:                                                                                                                                                                                          KDA004_HEPG2_96H:TRCN0000061769:-666|KDA004_HEPG2_96H:TRCN0000061770:-666|KDA004_HEPG2_96H:TRCN0000061772:-666
##  3:                                                                                                                                                                                          KDB010_HEPG2_96H:TRCN0000128018:-666|KDB010_HEPG2_96H:TRCN0000129686:-666|KDB010_HEPG2_96H:TRCN0000131227:-666
##  4:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000116523:-666|KDB009_HEPG2_96H:TRCN0000116524:-666|KDB009_HEPG2_96H:TRCN0000116525:-666
##  5:                                                                                                                                                                                                                               KDA003_HEPG2_96H:TRCN0000021241:-666|KDA003_HEPG2_96H:TRCN0000021242:-666
##  6:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000075318:-666|KDB008_HEPG2_96H:TRCN0000075320:-666|KDB008_HEPG2_96H:TRCN0000075321:-666
##  7:                                                                                                                                                                                                                               KDD006_HEPG2_96H:TRCN0000049389:-666|KDD006_HEPG2_96H:TRCN0000049390:-666
##  8:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000051813:-666|KDB008_HEPG2_96H:TRCN0000051815:-666|KDB008_HEPG2_96H:TRCN0000051816:-666
##  9:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000039004:-666|KDB009_HEPG2_96H:TRCN0000039005:-666|KDB009_HEPG2_96H:TRCN0000039007:-666
## 10: KDA005_HEPG2_96H:TRCN0000004265:-666|KDB007_HEPG2_96H:TRCN0000004265:-666|KDD006_HEPG2_96H:TRCN0000004265:-666|KDA005_HEPG2_96H:TRCN0000004266:-666|KDB007_HEPG2_96H:TRCN0000004266:-666|KDD006_HEPG2_96H:TRCN0000004266:-666|KDA005_HEPG2_96H:TRCN0000004267:-666|KDB007_HEPG2_96H:TRCN0000004267:-666</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get table with all information available (including ranks, metadata, compound</span>
<span class="co"># information, etc.)</span>
<span class="va">table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">as.table</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">table</span><span class="op">)</span></code></pre></div>
<pre><code>##                gene_perturbation spearman_coef pearson_coef      GSEA
## 1:   CGS001_HEPG2_96H:EIF4G1:1.5     0.1770993    0.1915142 0.4861485
## 2:    CGS001_HEPG2_96H:MECP2:1.5     0.1930335    0.1756788 0.3903862
## 3:   CGS001_HEPG2_96H:SKIV2L:1.5     0.1423207    0.1755377 0.5070024
## 4: CGS001_HEPG2_96H:KIAA0196:1.5     0.1894301    0.1847352 0.3521034
## 5:    CGS001_HEPG2_96H:STAT1:1.5     0.1833382    0.1894661 0.3765351
## 6:     CGS001_HEPG2_96H:MEST:1.5     0.1726479    0.1858475 0.4219693
##    spearman_rank pearson_rank GSEA_rank rankProduct_rank     pert_id pert_iname
## 1:             7            1         2                1 CGS001-1981     EIF4G1
## 2:             1            7        12                2 CGS001-4204      MECP2
## 3:            11            8         1                3 CGS001-6499     SKIV2L
## 4:             2            4        14                4 CGS001-9897   KIAA0196
## 5:             5            2        13                5 CGS001-6772      STAT1
## 6:             8            3        10                6 CGS001-4232       MEST
##     pert_type cell_id pert_idose pert_itime pr_gene_id
## 1: trt_sh.cgs   HEPG2 1.5 microL       96 h       1981
## 2: trt_sh.cgs   HEPG2 1.5 microL       96 h       4204
## 3: trt_sh.cgs   HEPG2 1.5 microL       96 h       6499
## 4: trt_sh.cgs   HEPG2 1.5 microL       96 h       9897
## 5: trt_sh.cgs   HEPG2 1.5 microL       96 h       6772
## 6: trt_sh.cgs   HEPG2 1.5 microL       96 h       4232
##                                          pr_gene_title pr_is_lm pr_is_bing
## 1: eukaryotic translation initiation factor 4 gamma, 1        1          1
## 2:                        methyl-CpG binding protein 2        0          1
## 3:                              Ski2 like RNA helicase        1          1
## 4:                                            KIAA0196        1          1
## 5:  signal transducer and activator of transcription 1        1          1
## 6:                        mesoderm specific transcript        1          1</code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Obtain available raw information from compared perturbations</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html">attributes</a></span><span class="op">(</span><span class="va">compareKD</span><span class="op">)</span><span class="op">)</span> <span class="co"># Data available in compared perturbations</span></code></pre></div>
<pre><code>##  [1] "class"             "row.names"         ".internal.selfref"
##  [4] "names"             "metadata"          "geneInfo"         
##  [7] "input"             "rankingInfo"       "geneset"          
## [10] "runtime"</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"metadata"</span><span class="op">)</span>  <span class="co"># Perturbation metadata</span></code></pre></div>
<pre><code>##                            sig_id      pert_id pert_iname  pert_type cell_id
##  1:    CGS001_HEPG2_96H:CDCA8:1.5 CGS001-55143      CDCA8 trt_sh.cgs   HEPG2
##  2:    CGS001_HEPG2_96H:COPS5:1.5 CGS001-10987      COPS5 trt_sh.cgs   HEPG2
##  3:    CGS001_HEPG2_96H:DHX16:1.5  CGS001-8449      DHX16 trt_sh.cgs   HEPG2
##  4:    CGS001_HEPG2_96H:EHMT2:1.5 CGS001-10919      EHMT2 trt_sh.cgs   HEPG2
##  5:   CGS001_HEPG2_96H:EIF4G1:1.5  CGS001-1981     EIF4G1 trt_sh.cgs   HEPG2
##  6:     CGS001_HEPG2_96H:EYA1:1.5  CGS001-2138       EYA1 trt_sh.cgs   HEPG2
##  7:    CGS001_HEPG2_96H:GNAI2:1.5  CGS001-2771      GNAI2 trt_sh.cgs   HEPG2
##  8:   CGS001_HEPG2_96H:GTPBP8:1.5 CGS001-29083     GTPBP8 trt_sh.cgs   HEPG2
##  9:      CGS001_HEPG2_96H:HFE:1.5  CGS001-3077        HFE trt_sh.cgs   HEPG2
## 10: CGS001_HEPG2_96H:KIAA0196:1.5  CGS001-9897   KIAA0196 trt_sh.cgs   HEPG2
## 11:   CGS001_HEPG2_96H:KIF20A:1.5 CGS001-10112     KIF20A trt_sh.cgs   HEPG2
## 12:      CGS001_HEPG2_96H:MAF:1.5  CGS001-4094        MAF trt_sh.cgs   HEPG2
## 13:    CGS001_HEPG2_96H:MECP2:1.5  CGS001-4204      MECP2 trt_sh.cgs   HEPG2
## 14:     CGS001_HEPG2_96H:MEST:1.5  CGS001-4232       MEST trt_sh.cgs   HEPG2
## 15:   CGS001_HEPG2_96H:NDUFB6:1.5  CGS001-4712     NDUFB6 trt_sh.cgs   HEPG2
## 16:    CGS001_HEPG2_96H:PLOD2:1.5  CGS001-5352      PLOD2 trt_sh.cgs   HEPG2
## 17:     CGS001_HEPG2_96H:PPIH:1.5 CGS001-10465       PPIH trt_sh.cgs   HEPG2
## 18:      CGS001_HEPG2_96H:SHB:1.5  CGS001-6461        SHB trt_sh.cgs   HEPG2
## 19:    CGS001_HEPG2_96H:SIAH2:1.5  CGS001-6478      SIAH2 trt_sh.cgs   HEPG2
## 20:   CGS001_HEPG2_96H:SKIV2L:1.5  CGS001-6499     SKIV2L trt_sh.cgs   HEPG2
## 21:    CGS001_HEPG2_96H:SQRDL:1.5 CGS001-58472      SQRDL trt_sh.cgs   HEPG2
## 22:    CGS001_HEPG2_96H:STAT1:1.5  CGS001-6772      STAT1 trt_sh.cgs   HEPG2
## 23:  CGS001_HEPG2_96H:SULT1A2:1.5  CGS001-6799    SULT1A2 trt_sh.cgs   HEPG2
## 24:    CGS001_HEPG2_96H:TMEM5:1.5 CGS001-10329      TMEM5 trt_sh.cgs   HEPG2
## 25:   CGS001_HEPG2_96H:UBAP2L:1.5  CGS001-9898     UBAP2L trt_sh.cgs   HEPG2
## 26:   CGS001_HEPG2_96H:ZBTB24:1.5  CGS001-9841     ZBTB24 trt_sh.cgs   HEPG2
##                            sig_id      pert_id pert_iname  pert_type cell_id
##     pert_dose pert_dose_unit pert_idose pert_time pert_time_unit pert_itime
##  1:       1.5         microL 1.5 microL        96              h       96 h
##  2:       1.5         microL 1.5 microL        96              h       96 h
##  3:       1.5         microL 1.5 microL        96              h       96 h
##  4:       1.5         microL 1.5 microL        96              h       96 h
##  5:       1.5         microL 1.5 microL        96              h       96 h
##  6:       1.5         microL 1.5 microL        96              h       96 h
##  7:       1.5         microL 1.5 microL        96              h       96 h
##  8:       1.5         microL 1.5 microL        96              h       96 h
##  9:       1.5         microL 1.5 microL        96              h       96 h
## 10:       1.5         microL 1.5 microL        96              h       96 h
## 11:       1.5         microL 1.5 microL        96              h       96 h
## 12:       1.5         microL 1.5 microL        96              h       96 h
## 13:       1.5         microL 1.5 microL        96              h       96 h
## 14:       1.5         microL 1.5 microL        96              h       96 h
## 15:       1.5         microL 1.5 microL        96              h       96 h
## 16:       1.5         microL 1.5 microL        96              h       96 h
## 17:       1.5         microL 1.5 microL        96              h       96 h
## 18:       1.5         microL 1.5 microL        96              h       96 h
## 19:       1.5         microL 1.5 microL        96              h       96 h
## 20:       1.5         microL 1.5 microL        96              h       96 h
## 21:       1.5         microL 1.5 microL        96              h       96 h
## 22:       1.5         microL 1.5 microL        96              h       96 h
## 23:       1.5         microL 1.5 microL        96              h       96 h
## 24:       1.5         microL 1.5 microL        96              h       96 h
## 25:       1.5         microL 1.5 microL        96              h       96 h
## 26:       1.5         microL 1.5 microL        96              h       96 h
##     pert_dose pert_dose_unit pert_idose pert_time pert_time_unit pert_itime
##                                                                                                                                                                                                                                                                                                   distil_id
##  1:                                                                                                                                                                                          KDA003_HEPG2_96H:TRCN0000007898:-666|KDA003_HEPG2_96H:TRCN0000007899:-666|KDA003_HEPG2_96H:TRCN0000007900:-666
##  2:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000019200:-666|KDB009_HEPG2_96H:TRCN0000019201:-666|KDB009_HEPG2_96H:TRCN0000019202:-666
##  3:                                                                                                                                                                                          KDB010_HEPG2_96H:TRCN0000001123:-666|KDB010_HEPG2_96H:TRCN0000001124:-666|KDB010_HEPG2_96H:TRCN0000001125:-666
##  4:                                                                                                                                                                                          KDA002_HEPG2_96H:TRCN0000115667:-666|KDA002_HEPG2_96H:TRCN0000115669:-666|KDA002_HEPG2_96H:TRCN0000115671:-666
##  5:                                                                                                                                                                                          KDA004_HEPG2_96H:TRCN0000061769:-666|KDA004_HEPG2_96H:TRCN0000061770:-666|KDA004_HEPG2_96H:TRCN0000061772:-666
##  6:                                                                                                                                                                                          KDC008_HEPG2_96H:TRCN0000083443:-666|KDC008_HEPG2_96H:TRCN0000083446:-666|KDC008_HEPG2_96H:TRCN0000303461:-666
##  7:                                                                           KDA006_HEPG2_96H:TRCN0000029514:-666|KDB003_HEPG2_96H:TRCN0000029514:-666|KDA006_HEPG2_96H:TRCN0000029517:-666|KDB003_HEPG2_96H:TRCN0000029517:-666|KDA006_HEPG2_96H:TRCN0000278333:-666|KDB003_HEPG2_96H:TRCN0000029518:-666
##  8:                                                                                                                                                                                          KDB001_HEPG2_96H:TRCN0000343725:-666|KDB001_HEPG2_96H:TRCN0000343727:-666|KDB001_HEPG2_96H:TRCN0000343728:-666
##  9:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000060018:-666|KDB008_HEPG2_96H:TRCN0000060019:-666|KDB008_HEPG2_96H:TRCN0000060022:-666
## 10:                                                                                                                                                                                          KDB010_HEPG2_96H:TRCN0000128018:-666|KDB010_HEPG2_96H:TRCN0000129686:-666|KDB010_HEPG2_96H:TRCN0000131227:-666
## 11:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000116523:-666|KDB009_HEPG2_96H:TRCN0000116524:-666|KDB009_HEPG2_96H:TRCN0000116525:-666
## 12:                                                                                                                                                                                          KDC001_HEPG2_96H:TRCN0000000254:-666|KDC001_HEPG2_96H:TRCN0000000256:-666|KDC001_HEPG2_96H:TRCN0000000258:-666
## 13:                                                                                                                                                                                                                               KDA003_HEPG2_96H:TRCN0000021241:-666|KDA003_HEPG2_96H:TRCN0000021242:-666
## 14:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000075318:-666|KDB008_HEPG2_96H:TRCN0000075320:-666|KDB008_HEPG2_96H:TRCN0000075321:-666
## 15:                                                                                                                                                                                          KDB010_HEPG2_96H:TRCN0000028409:-666|KDB010_HEPG2_96H:TRCN0000028417:-666|KDB010_HEPG2_96H:TRCN0000028428:-666
## 16:                                                                                                                                                                                          KDC010_HEPG2_96H:TRCN0000064808:-666|KDC010_HEPG2_96H:TRCN0000064809:-666|KDC010_HEPG2_96H:TRCN0000064810:-666
## 17:                                                                                                                                                                                                                               KDD006_HEPG2_96H:TRCN0000049389:-666|KDD006_HEPG2_96H:TRCN0000049390:-666
## 18:                                                                                                                                                                                          KDB001_HEPG2_96H:TRCN0000149080:-666|KDB001_HEPG2_96H:TRCN0000282026:-666|KDB001_HEPG2_96H:TRCN0000285376:-666
## 19:                                                                                                                                                                                          KDA005_HEPG2_96H:TRCN0000007415:-666|KDA005_HEPG2_96H:TRCN0000007416:-666|KDA005_HEPG2_96H:TRCN0000007418:-666
## 20:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000051813:-666|KDB008_HEPG2_96H:TRCN0000051815:-666|KDB008_HEPG2_96H:TRCN0000051816:-666
## 21:                                                                                                                                                                                          KDB009_HEPG2_96H:TRCN0000039004:-666|KDB009_HEPG2_96H:TRCN0000039005:-666|KDB009_HEPG2_96H:TRCN0000039007:-666
## 22: KDA005_HEPG2_96H:TRCN0000004265:-666|KDB007_HEPG2_96H:TRCN0000004265:-666|KDD006_HEPG2_96H:TRCN0000004265:-666|KDA005_HEPG2_96H:TRCN0000004266:-666|KDB007_HEPG2_96H:TRCN0000004266:-666|KDD006_HEPG2_96H:TRCN0000004266:-666|KDA005_HEPG2_96H:TRCN0000004267:-666|KDB007_HEPG2_96H:TRCN0000004267:-666
## 23:                                                                                                                                                                                          KDA001_HEPG2_96H:TRCN0000035200:-666|KDA001_HEPG2_96H:TRCN0000035202:-666|KDA001_HEPG2_96H:TRCN0000035203:-666
## 24:                                                                                                                                                                                                                               KDB001_HEPG2_96H:TRCN0000135504:-666|KDB001_HEPG2_96H:TRCN0000278433:-666
## 25:                                                                                                                                                                                          KDA003_HEPG2_96H:TRCN0000007677:-666|KDA003_HEPG2_96H:TRCN0000007680:-666|KDA003_HEPG2_96H:TRCN0000007681:-666
## 26:                                                                                                                                                                                          KDB008_HEPG2_96H:TRCN0000107435:-666|KDB008_HEPG2_96H:TRCN0000107437:-666|KDB008_HEPG2_96H:TRCN0000107438:-666
##                                                                                                                                                                                                                                                                                                   distil_id</code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="st">"geneInfo"</span><span class="op">)</span>  <span class="co"># Gene information</span></code></pre></div>
<pre><code>##        pr_gene_id pr_gene_symbol
##     1:        780           DDR1
##     2:       7849           PAX8
##     3:       2978         GUCA1A
##     4:       2049          EPHB3
##     5:       2101          ESRRA
##    ---                          
## 12324:       4034          LRCH4
## 12325:     399664          MEX3D
## 12326:      54869         EPS8L1
## 12327:      90379         DCAF15
## 12328:         60           ACTB
##                                                              pr_gene_title
##     1:                         discoidin domain receptor tyrosine kinase 1
##     2:                                                        paired box 8
##     3:                                      guanylate cyclase activator 1A
##     4:                                                     EPH receptor B3
##     5:                                     estrogen related receptor alpha
##    ---                                                                    
## 12324: leucine-rich repeats and calponin homology (CH) domain containing 4
## 12325:                                   mex-3 RNA binding family member D
## 12326:                                                         EPS8 like 1
## 12327:                                  DDB1 and CUL4 associated factor 15
## 12328:                                                         actin, beta
##        pr_is_lm pr_is_bing
##     1:        1          1
##     2:        1          1
##     3:        0          0
##     4:        0          1
##     5:        0          1
##    ---                    
## 12324:        0          1
## 12325:        0          1
## 12326:        0          1
## 12327:        0          1
## 12328:        0          1</code></pre>
</div>
<div id="relationship-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#relationship-plots" class="anchor"></a>Relationship plots</h1>
<p>To analyse the relationship between the user-provided differential expression profile with that associated with a specific perturbation, scatter plots (for Spearman and Pearson analyses) and GSEA plots are available.</p>
<p>For instance, let’s plot the relationship between EIF4G1 shRNA knockdown from ENCODE with the CMap knockdown perturbations:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot relationship with EIF4G1 knockdown from CMap</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">EIF4G1knockdown</span>, <span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-1.png" width="480"></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">EIF4G1knockdown</span>, <span class="st">"pearson"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-2.png" width="480"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">EIF4G1knockdown</span>, <span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-3.png" width="480"></p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot relationship with most negatively associated perturbation</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">spearman_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-4.png" width="480"></p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">pearson_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"pearson"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-5.png" width="480"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareKD</span>, <span class="va">compareKD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">GSEA_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20knockdown%20comparison-6.png" width="480"></p>
<p>For small molecules:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot relationship with most positively associated perturbation</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">spearman_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-1.png" width="480"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">pearson_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"pearson"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-2.png" width="480"></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">GSEA_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-3.png" width="480"></p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot relationship with most negatively associated perturbation</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">spearman_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span>, <span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-4.png" width="480"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">pearson_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"pearson"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-5.png" width="480"></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">compareCompounds</span>, <span class="va">compareCompounds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">GSEA_rank</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/plot%20small%20molecule%20comparison-6.png" width="480"></p>
</div>
<div id="predict-targeting-drugs" class="section level1">
<h1 class="hasAnchor">
<a href="#predict-targeting-drugs" class="anchor"></a>Predict targeting drugs</h1>
<p>Compounds that target the phenotypes associated with the user-provided differential expression profile can be inferred by comparing against gene expression and drug sensitivity associations. The gene expression and drug sensitivity datasets derived from the following sources were correlated using Spearman’s correlation across the available cell lines.</p>
<table class="table">
<thead><tr class="header">
<th>Source</th>
<th align="right">Screened compounds</th>
<th align="right">Human cancer cell lines</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://dtp.cancer.gov/discovery_development/nci-60">NCI60</a></td>
<td align="right">&gt; 100 000</td>
<td align="right">60</td>
</tr>
<tr class="even">
<td><a href="https://www.cancerrxgene.org">GDSC 7</a></td>
<td align="right">481</td>
<td align="right">860</td>
</tr>
<tr class="odd">
<td><a href="https://portals.broadinstitute.org/ctrp/">CTRP 2.1</a></td>
<td align="right">138</td>
<td align="right">~700</td>
</tr>
</tbody>
</table>
<p>To use an expression and drug sensitivity association based on <code>CTRP 2.1</code> (<code>GDSC 7</code> and <code>NCI60</code> could be used instead) to infer targeting drugs for the user’s differential expression profile:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/listExpressionDrugSensitivityAssociation.html">listExpressionDrugSensitivityAssociation</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "GDSC 7"   "CTRP 2.1" "NCI60"</code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ctrp</span>      <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listExpressionDrugSensitivityAssociation.html">listExpressionDrugSensitivityAssociation</a></span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">assoc</span>     <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadExpressionDrugSensitivityAssociation.html">loadExpressionDrugSensitivityAssociation</a></span><span class="op">(</span><span class="va">ctrp</span><span class="op">)</span></code></pre></div>
<pre><code>## expressionDrugSensitivityCorCTRP2.1.rds not found: downloading data...</code></pre>
<pre><code>## Loading data from expressionDrugSensitivityCorCTRP2.1.rds...</code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predicted</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictTargetingDrugs.html">predictTargetingDrugs</a></span><span class="op">(</span><span class="va">diffExprStat</span>, <span class="va">assoc</span><span class="op">)</span></code></pre></div>
<pre><code>## Subsetting data based on 11227 intersecting genes (83% of the 13451 input genes)...</code></pre>
<pre><code>## Comparing against 545 CTRP 2.1 compounds (823 cell lines) using 'spearman, pearson, gsea' (gene size of 150)...</code></pre>
<pre><code>## Comparison performed in 8.46 secs</code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">predicted</span>, method<span class="op">=</span><span class="st">"rankProduct"</span><span class="op">)</span></code></pre></div>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot results for a given drug</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">predicted</span>, <span class="va">predicted</span><span class="op">[[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">]</span>, method<span class="op">=</span><span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading data from /Users/runner/work/cTRAP/cTRAP/vignettes/expressionDrugSensitivityCorCTRP2.1.rds...</code></pre>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-4-2.png" width="480"></p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">predicted</span>, <span class="va">predicted</span><span class="op">[[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">]</span>, method<span class="op">=</span><span class="st">"gsea"</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading data from /Users/runner/work/cTRAP/cTRAP/vignettes/expressionDrugSensitivityCorCTRP2.1.rds...</code></pre>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-4-3.png" width="480"></p>
<blockquote>
<p>Compounds are ranked by their relative targeting potential based on the input differential expression profile (i.e. the 1st-ranked compound has higher targeting potential than the 2nd-ranked one).</p>
</blockquote>
<p>Candidate targeting drugs can be plotted against the similarity ranking of their perturbations towards the user’s differential expression profile. Note that the highlighted values are the same compounds for the following plots annotated with their name, gene target and mechanism of action (MOA), respectively.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Label by compound name</span>
<span class="fu"><a href="../reference/plotTargetingDrugsVSsimilarPerturbations.html">plotTargetingDrugsVSsimilarPerturbations</a></span><span class="op">(</span>
  <span class="va">predicted</span>, <span class="va">compareCompounds</span>, column<span class="op">=</span><span class="st">"spearman_rank"</span><span class="op">)</span></code></pre></div>
<pre><code>## Columns 'broad id' and 'pubchem_cid' were matched based on 0 common values; to manually select columns to compare, please set arguments starting with 'keyCol'</code></pre>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Label by compound's gene target</span>
<span class="fu"><a href="../reference/plotTargetingDrugsVSsimilarPerturbations.html">plotTargetingDrugsVSsimilarPerturbations</a></span><span class="op">(</span>
  <span class="va">predicted</span>, <span class="va">compareCompounds</span>, column<span class="op">=</span><span class="st">"spearman_rank"</span>, labelBy<span class="op">=</span><span class="st">"target"</span><span class="op">)</span></code></pre></div>
<pre><code>## Columns 'broad id' and 'pubchem_cid' were matched based on 0 common values; to manually select columns to compare, please set arguments starting with 'keyCol'</code></pre>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-5-2.png" width="480"></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Label by compound's mechanism of action (MOA)</span>
<span class="fu"><a href="../reference/plotTargetingDrugsVSsimilarPerturbations.html">plotTargetingDrugsVSsimilarPerturbations</a></span><span class="op">(</span>
  <span class="va">predicted</span>, <span class="va">compareCompounds</span>, column<span class="op">=</span><span class="st">"spearman_rank"</span>, labelBy<span class="op">=</span><span class="st">"moa"</span><span class="op">)</span></code></pre></div>
<pre><code>## Columns 'broad id' and 'pubchem_cid' were matched based on 0 common values; to manually select columns to compare, please set arguments starting with 'keyCol'</code></pre>
<p><img src="cTRAP_files/figure-html/unnamed-chunk-5-3.png" width="480"></p>
</div>
<div id="molecular-descriptor-enrichment-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#molecular-descriptor-enrichment-analysis" class="anchor"></a>Molecular descriptor enrichment analysis</h1>
<p>Next, from our candidate targeting drugs, we will analyse the enrichment of 2D and 3D molecular descriptors based on CMap and NCI60 compounds. This allows to check if targeting drugs are particularly enriched in specific drug descriptors and may be useful to think about the relevance of these descriptors for targeting a phenotype of interest.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">descriptors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadDrugDescriptors.html">loadDrugDescriptors</a></span><span class="op">(</span><span class="st">"CMap"</span>, <span class="st">"2D"</span><span class="op">)</span></code></pre></div>
<pre><code>## molecular_descriptors_CMap_2D.rds not found: downloading data...</code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">drugSets</span>    <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareDrugSets.html">prepareDrugSets</a></span><span class="op">(</span><span class="va">descriptors</span><span class="op">)</span>
<span class="va">dsea</span>        <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyseDrugSetEnrichment.html">analyseDrugSetEnrichment</a></span><span class="op">(</span><span class="va">drugSets</span>, <span class="va">predicted</span><span class="op">)</span></code></pre></div>
<pre><code>## Matching compounds with those available in drug sets...</code></pre>
<pre><code>## Ordering results by column 'rankProduct_rank'; to manually select column to order by, please set argument 'col'</code></pre>
<pre><code>## Columns 'broad id' and 'pert_id' were matched based on 283 common values; to manually select columns to compare, please set arguments starting with 'keyCol'</code></pre>
<pre><code>## Performing enrichment analysis...</code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot the 6 most significant drug set enrichment results</span>
<span class="fu"><a href="../reference/plotDrugSetEnrichment.html">plotDrugSetEnrichment</a></span><span class="op">(</span><span class="va">drugSets</span>, <span class="va">predicted</span>,
                      selectedSets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dsea</span><span class="op">$</span><span class="va">descriptor</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Matching compounds with those available in drug sets...</code></pre>
<pre><code>## Columns 'broad id' and 'pert_id' were matched based on 283 common values; to manually select columns to compare, please set arguments starting with 'keyCol'</code></pre>
<pre><code>## Plotting enrichment analysis...</code></pre>
<pre><code>## $`Total Surface Area: [293, 307]`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-1.png" width="480"></p>
<pre><code>## 
## $`Rotatable Bonds: 5`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-2.png" width="480"></p>
<pre><code>## 
## $`Total.Molweight: [391, 411]`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-3.png" width="480"></p>
<pre><code>## 
## $`Polar.Surface.Area: [138, 1153]`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-4.png" width="480"></p>
<pre><code>## 
## $`Total Molweight: [391, 411]`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-5.png" width="480"></p>
<pre><code>## 
## $`cLogP: 0`</code></pre>
<p><img src="cTRAP_files/figure-html/drug%20set%20enrichment%20analysis-6.png" width="480"></p>
</div>
<div id="contact-information" class="section level1">
<h1 class="hasAnchor">
<a href="#contact-information" class="anchor"></a>Contact information</h1>
<p>All feedback on the program, documentation and associated material (including this tutorial) is welcome. Please send any suggestions and comments to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt" class="email">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p>Bernardo P. de Almeida (<a href="mailto:bernardo.almeida94@gmail.com" class="email">bernardo.almeida94@gmail.com</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Bernardo P. de Almeida, Nuno Saraiva-Agostinho, Nuno L. Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
